File: backend-v5/lms/projects/Project.php
<?php

// Modul: lms/projects
// Klasa: Project
// Opis: jedan projekat / seminarski rad

require_once(Config::$backend_path."core/DB.php");
require_once(Config::$backend_path."core/Person.php");

class Project {
	public $id;
	public $name, $courseUnitId, $academicYearId, $description, $time, $note;
	
	public static function fromId($id) {
		$q10 = DB::query("select naziv, predmet, akademska_godina, opis, UNIX_TIMESTAMP(vrijeme), biljeska from projekat where id=$id");
		if (mysql_num_rows($q10)<1) {
			throw new Exception("unknown project");
		}
		
		$p = new Project;
		$p->id = $id;
		$p->name = mysql_result($q10,0,0);
		$p->courseUnitId = mysql_result($q10,0,1);
		$p->academicYearId = mysql_result($q10,0,2);
		$p->description = mysql_result($q10,0,3);
		$p->time = mysql_result($q10,0,4);
		$p->note = mysql_result($q10,0,5);
		
		return $p;
	}
	
	public static function getAllForCourse($courseUnitId, $academicYearId) {
		$q10 = DB::query("select id, naziv, opis, UNIX_TIMESTAMP(vrijeme), biljeska from projekat where predmet=$courseUnitId and akademska_godina=$academicYearId order by naziv");
		$projects = array();
		while ($r10 = mysql_fetch_row($q10)) {
			$p = new Project;
			$p->id = $r10[0];
			$p->name = $r10[1];
			$p->courseUnitId = $courseUnitId;
			$p->academicYearId = $academicYearId;
			$p->description = $r10[2];
			$p->time = $r10[3];
			$p->note = $r10[4];

			array_push($projects, $p);
		}
		
		return $projects;
	}
	
	public static function fromMemberAndCourse($personId, $courseUnitId, $academicYearId) {
		$q10 = DB::query("select p.id, p.naziv, p.opis, UNIX_TIMESTAMP(p.vrijeme), p.biljeska from projekat as p, student_projekat as sp where p.predmet=$courseUnitId and p.akademska_godina=$academicYearId and p.id=sp.projekat and sp.student=$personId");
		if (mysql_num_rows($q10)<1) {
			throw new Exception("person not a member of any project in this course");
		}
		
		$p = new Project;
		$p->id = $id;
		$p->name = mysql_result($q10,0,0);
		$p->courseUnitId = $courseUnitId;
		$p->academicYearId = $academicYearId;
		$p->description = mysql_result($q10,0,1);
		$p->time = mysql_result($q10,0,2);
		$p->note = mysql_result($q10,0,3);
		
		return $p;
	}

	public function isMember($personId) {
		$q10 = DB::query("select count(*) from student_projekat where student=$personId and projekat=".$this->id);
		if (mysql_result($q10,0,0)>0) return true;
		return false;
	}

	public function getMembers() {
		$q10 = DB::query("select o.id, o.ime, o.prezime, o.brindexa from student_projekat as sp, osoba as o where sp.projekat=".$this->id." and sp.student=o.id");
		$members = array();
		while ($r10 = mysql_fetch_row($q10)) {
			$p = new Person;
			$p->id = $r10[0];
			$p->name = $r10[1];
			$p->surname = $r10[2];
			$p->studentIdNr = $r10[3];
			$p->login = ""; // FIXME: problem je što može postojati više slogova u tabeli auth za jedan login a hoću da izbjegnem podupite

			array_push($members, $p);
		}
		return $members;
	}

	public function addMember($personId) {
		$q10 = DB::query("INSERT INTO student_projekat SET student=$personId, projekat=".$this->id);
	}

	public function deleteMember($personId) {
		$q10 = DB::query("DELETE FROM student_projekat WHERE student=$personId AND projekat=".$this->id);
	}
}

?>

File: rorbackend/app/models/lms/project/project.rb
class Lms::Project::Project < ActiveRecord::Base
  # Uncomment following lines if working with legacy database
  # set_table_name 'projekat'
  # set_primary_key :id
  # alias_attribute :id, :id
  # alias_attribute :name, :naziv
  # alias_attribute :course_unit_id, :predmet
  # alias_attribute :academic_year_id, :akademska_godina
  # alias_attribute :description, :opis
  # alias_attribute :note, :biljeska
  # alias_attribute :time, :vrijeme

  # Uncomment following lines if working with legacy database
  # TABLE_NAME = 'projekat'
  # ID = TABLE_NAME + '.' + 'id'
  # NAME = TABLE_NAME + '.' + 'naziv'
  # COURSE_UNIT_ID = TABLE_NAME + '.' + 'predmet'
  # ACADEMIC_YEAR_ID = TABLE_NAME + '.' + 'akademska_godina'
  # DESCRIPTION = TABLE_NAME + '.' + 'opis'
  # NOTE = TABLE_NAME + '.' + 'biljeska'
  # TIME = TABLE_NAME + '.' + 'vrijeme'

  # Comment following lines if working with legacy database
  TABLE_NAME = 'lms_project_projects'
  ID = TABLE_NAME + '.' + 'id'
  NAME = TABLE_NAME + '.' + 'name'
  COURSE_UNIT_ID = TABLE_NAME + '.' + 'course_unit_id'
  ACADEMIC_YEAR_ID = TABLE_NAME + '.' + 'academic_year_id'
  DESCRIPTION = TABLE_NAME + '.' + 'description'
  NOTE = TABLE_NAME + '.' + 'note'
  TIME = TABLE_NAME + '.' + 'time'

  ALL_COLUMNS = [ID, COURSE_UNIT_ID, ACADEMIC_YEAR_ID, DESCRIPTION, NOTE, TIME]
  
  belongs_to :course_unit, :class_name => "Core::CourseUnit"
  belongs_to :academic_year, :class_name => "Core::AcademicYear"
  has_many :project_students
  
  
  def self.from_member_and_course(course_unit_id, academic_year_id)
    project = (Lms::Project::Project).joins(:project_students).where(:course_unit_id => course_unit_id, :academic_year_id => academic_year_id, (Lms::Project::ProjectStudent).STUDENT_ID => student_id).first
    return project
  end
  
  
  def self.get_all_for_course(course_unit_id)
    pojects = (Lms::Project::Project).where(:course_unit_id => course_unit_id)
    return projects
  end
  
  
  def self.is_member(id, student_id)
    num_results = (Lms::Project::ProjectStudent).where(:student_id => student_id, :project_id => id).count
    member = true
    if num_results == 0
      member = false
    end
    return member
  end
  
  
  def self.get_members(id)
    select_columns = [(Core::People).ID, (Core::People).NAME, (Core::People).SURNAME, (Core::People).STUDENT_ID_NUMBER]
    members = (Lms::Project::ProjectStudent).joins(:student).where(:project_id => id).select(select_columns)
    return members
  end
  
  
  def self.add_member(id, student_id)
    project_student = (Lms::Project::ProjectStudent).new(:student_id => student_id, :project_id => id)
    return project_student.save
  end
  
  
  def self.delete_member(id, student_id)
    member = (Lms::Project::ProjectStudent).delete_all([:student_id => student_id, :project_id => id])
    raise ActiveRecord::RecordNotFound if member == 0
    return true
  end
  
  
end
